src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -132,55 +132,69 @@ private async Task WaitUntilHealthyAsync(IResource resource, IResource dependenc
     {
         var resourceLogger = _resourceLoggerService.GetLogger(resource);
         resourceLogger.LogInformation("Waiting for resource '{Name}' to enter the '{State}' state.", dependency.Name, KnownResourceStates.Running);
-
         await PublishUpdateAsync(resource, s => s with { State = KnownResourceStates.Waiting }).ConfigureAwait(false);
-        var resourceEvent = await WaitForResourceCoreAsync(dependency.Name, re => IsContinuableState(re.Snapshot), cancellationToken: cancellationToken).ConfigureAwait(false);
-        var snapshot = resourceEvent.Snapshot;
 
-        if (snapshot.State?.Text == KnownResourceStates.FailedToStart)
-        {
-            resourceLogger.LogError(
-                "Dependency resource '{ResourceName}' failed to start.",
-                dependency.Name
-                );
+        var names = dependency.GetResolvedResourceNames();
+        var tasks = new Task[names.Length];
 
-            throw new DistributedApplicationException($"Dependency resource '{dependency.Name}' failed to start.");
-        }
-        else if (snapshot.State!.Text == KnownResourceStates.Finished || snapshot.State!.Text == KnownResourceStates.Exited)
+        for (var i = 0; i < names.Length; i++)
         {
-            resourceLogger.LogError(
-                "Resource '{ResourceName}' has entered the '{State}' state prematurely.",
-                dependency.Name,
-                snapshot.State.Text
-                );
-
-            throw new DistributedApplicationException(
-                $"Resource '{dependency.Name}' has entered the '{snapshot.State.Text}' state prematurely."
-                );
+            var displayName = names.Length > 1 ? names[i] : dependency.Name;
+            tasks[i] = Core(displayName, names[i]);
         }
 
-        // If our dependency resource has health check annotations we want to wait until they turn healthy
-        // otherwise we don't care about their health status.
-        if (dependency.TryGetAnnotationsOfType<HealthCheckAnnotation>(out var _))
+        await Task.WhenAll(tasks).ConfigureAwait(false);
+
+        async Task Core(string displayName, string resourceId)
         {
-            resourceLogger.LogInformation("Waiting for resource '{Name}' to become healthy.", dependency.Name);
-            await WaitForResourceHealthyAsync(dependency.Name, cancellationToken).ConfigureAwait(false);
-        }
+            var resourceEvent = await WaitForResourceCoreAsync(dependency.Name, re => re.ResourceId == resourceId && IsContinuableState(re.Snapshot), cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Why isn't this


```suggestion
            var resourceEvent = await WaitForResourceCoreAsync(displayName, re => re.ResourceId == resourceId && IsContinuableState(re.Snapshot), cancellationToken: cancellationToken).ConfigureAwait(false);
```

?

Reply:
Is it because it explicitly needs to be the Aspire Resource.Name?

Reply:
Even though they are the same, I want to keep display name for the display only.